rat/compiler.py

```diff
@@ -523,13 +587,59 @@ def _generate_python_source(self, ordered_expr_trees: List[ops.Expr]):
                 code += f"        {constrained_reference} = jax.numpy.pad({constrained_reference}, (0, {variable.padded_size() - size}))\n"
         code += "\n    return total, parameters\n"
 
+        ordered_expr_trees = list(reversed(self.expr_tree_list))
+
         # Transform parameters
         code += "\ndef transform_parameters(data, subscripts, parameters):"
         for top_expr in ordered_expr_trees:
-            if isinstance(top_expr, ops.Assignment):
-                lhs_key = top_expr.lhs.get_key()
-                code += f"\n    # {lhs_key}\n"
-                code += f"    parameters['{lhs_key}'] = {top_expr.rhs.code()}\n"
+            match top_expr:
+                case ops.Assignment(lhs, rhs):
+                    # Assume left hand side is a parameter
+                    lhs_key = top_expr.lhs.get_key()
+                    code += f"\n    # {lhs_key}\n"
+
+                    lhs_size = lhs.variable.size()
+                    lhs_is_scalar = lhs_size == None
+                    lhs_used_in_rhs = False
+                    if not lhs_is_scalar:
+                        carry_size = 0
+                        # Because left hand side is parameter, we only need to search for parameters on the right hand side
+                        for symbol in ops.search_tree(rhs, ops.Param):
+                            symbol_key = symbol.get_key()
+
+                            # Save an indicator if the left hand side is used on the right hand side
+                            if symbol_key == lhs_key:
+                                lhs_used_in_rhs = True
+
+                                # Figure out the carry size needed
+                                shifts = symbol.subscript.shifts
+
+                                number_subscripts_shifted = sum(shift is not None for shift in shifts)
+                                if number_subscripts_shifted != 1 or shifts[0] is None:
+                                    raise Exception(
+                                        "Internal compiler error (shouldn't alert here): When shifting an assigned variable, can only have one shifted subscript and it must be the first"
+                                    )
+
+                                carry_size = max(carry_size, shifts[0])
+
+                    if lhs_used_in_rhs:
```

Comment:
When we're using the variable assigned on the left hand side on the right hand side, we know we need to evaluate everything with a scan.

For the code:
```
y' ~ normal(mu[i], 0.1);
mu[i]' = mu[shift(i, 1)] + epsilon[i];
epsilon[i] ~ normal(0, 0.3);
```

We'll get a transformed parameters block like:
```python
ef transform_parameters(data, subscripts, parameters):
    # mu
    def scan_function_6(carry, index):
        next_value = carry[index - subscripts['i__1__3'][index] - 1] + parameters['epsilon'][subscripts['i__4'][index]]
        next_value_as_array = jax.numpy.array([next_value])
        return jax.numpy.concatenate([next_value_as_array, carry[1:]]), next_value

    _, parameters['mu'] = jax.lax.scan(scan_function_6, jax.numpy.zeros(1), jax.numpy.arange(10))

    return parameters
```